lib/ansible/vars/manager.py

```diff
@@ -451,7 +451,23 @@ def _get_magic_variables(self, play, host, task, include_hostvars, include_deleg
         variables['ansible_playbook_python'] = sys.executable
 
         if play:
-            variables['role_names'] = [r._role_name for r in play.roles]
+            # This is a list of all role names of all dependencies for all roles for this play
+            dependency_role_names = list(set([d._role_name for r in play.roles for d in r.get_all_dependencies()]))
+            # This is a list of all role names of all roles for this play
+            play_role_names = [r._role_name for r in play.roles]
+
+            # ansible_role_names includes all role names, dependent or directly referenced by the play
+            variables['ansible_role_names'] = list(set(dependency_role_names + play_role_names))
+            # ansible_play_role_names includes the names of all roles directly referenced by this play
+            # roles that are implicitly referenced via dependencies are not listed.
+            variables['ansible_play_role_names'] = play_role_names
+            # ansible_dependent_role_names includes the names of all roles that are referenced via dependencies
+            # dependencies that are also explicitly named as roles are included in this list
+            variables['ansible_dependent_role_names'] = dependency_role_names
+
+            # DEPRECATED: role_names should be deprecated in favor of ansible_role_names or ansible_play_role_names
+            variables['role_names'] = variables['ansible_play_role_names']
```

Comment:
as in previous comment, don't change `role_names` the new variable provides what you want, don't change the existing variable as people are using the current behavior.

Reply:
The functionality of role_names remains the same before- and after-change; I merely moved the primary point of assignment to `ansible_play_role_names` and made `role_names` a copy from that.

This behavior is a copy from lines 491-493: https://github.com/ansible/ansible/blob/884ff74353a327165b9f4afcdae5081d006173ec/lib/ansible/vars/manager.py#L491-L493

Reply:
ah, misread, thought you were making it same as ansible_role_names